[libs] E: creates config crate

danbugs opened the pull request:
closes #407

danbugs committed:
[libs] E: create new config crate to centralize config options

Signed-off-by: danbugs <danilochiarlone@gmail.com>

danbugs committed:
[libs] E: moved libs/sys/.../config to and constants to config crate and replaced usages

Signed-off-by: danbugs <danilochiarlone@gmail.com>

ppenna commented on src/kernel/src/mm/kheap.rs:
```diff
@@ -14,12 +14,12 @@ use ::alloc::alloc::{
         ErrorCode,
     },
 };
+use config::constants;
```
Could you please prefix it with `::` to keep consistency?

danbugs replied:
Done in [f588481](https://github.com/nanvix/nanvix/pull/434/commits/f58848116710b2cf3e7335e42815fc2cea5c5f1a) ✅

ppenna commented on src/kernel/src/mm/phys/frame.rs:
```diff
@@ -22,12 +22,12 @@ use ::sys::{
         ErrorCode,
     },
 };
+use config::constants;
```
Could you please prefix it with `::` to keep consistency?

danbugs replied:
Done in [f588481](https://github.com/nanvix/nanvix/pull/434/commits/f58848116710b2cf3e7335e42815fc2cea5c5f1a) ✅

ppenna reviewed (commented):
So far, so good!

danbugs committed:
[libs] E: adding '::' prefix to usage of config crate

Signed-off-by: danbugs <danilochiarlone@gmail.com>

danbugs committed:
[libs] E: moved hal config to new crate

- moved default values for: boot magic, initrd base, and stdout/stdin/vmm ports.
- cleaned up platforms to use default constants rather than literal values.

Signed-off-by: danbugs <danilochiarlone@gmail.com>

danbugs committed:
[libs] E: kernel config constants can now be read outside Rust

Created a .toml file w/ all kernel config constants. This .toml
file can be easily read for a shell script (e.g., run.sh) so that
we have a single source of truth for target configuration. The logic
to do this can be seen in the build.rs file.

Note: chose to use TOML vs. JSON to conserve the comments that
were present in code.

Signed-off-by: danbugs <danilochiarlone@gmail.com>

ppenna commented on src/libs/config/build.rs:
```diff
@@ -0,0 +1,48 @@
+
+fn main() {
+    // Read the TOML file
+    let kernel_config_content = fs::read_to_string("../../../config/kernel_config.toml").expect("Failed to read kernel_config.toml");
```
We can create a symlink to build directory and avoid this hardcode.

danbugs replied:
Done in [570a85c](https://github.com/nanvix/nanvix/pull/434/commits/570a85cd10c6775a7d73c26cda2546aa8a871414) ✅

ppenna commented on src/libs/config/build.rs:
```diff
@@ -0,0 +1,48 @@
+    fs::write(&dest_path, constants).expect("Failed to write kernel_config.rs");
+
+    // Inform Cargo to rerun the build script if the TOML changes
+    println!("cargo:rerun-if-changed=kernel_config.toml");
```
No need to provide full/reative path?

danbugs replied:
Fixed in [570a85c](https://github.com/nanvix/nanvix/pull/434/commits/570a85cd10c6775a7d73c26cda2546aa8a871414) ✅

ppenna commented on src/libs/config/src/lib.rs:
```diff
@@ -0,0 +1,73 @@
+//==================================================================================================
+// Hardware Abstraction Layer
+//==================================================================================================
+pub mod hal {
+    /// Magic value that identifies the virtual machine monitor.
+    pub const DEFAULT_BOOT_MAGIC: u32 = 0x0c00ffee;
+
+    /// Base address of the RAM disk.
+    pub const DEFAULT_INITRD_BASE: usize = 0x00800000;
+
+    /// I/O port that is connected to the standard output of the virtual machine.
+    pub const DEFAULT_STDOUT_PORT: u16 = 0xe9;
+
+    /// I/O port that is connected to the standard input of the virtual machine.
+    pub const DEFAULT_STDIN_PORT: u16 = 0xea;
+
+    /// I/O port that enables the guest to invoke functionalities of the virtual machine monitor.
+    pub const DEFAULT_VMM_PORT: u16 = 0x604;
+}
```
I would not move these here, as they are not config but constants (which are platform dependent)

danbugs replied:
I just noticed a lot of common usages for these constants hardcoded around the codebase. Figured it might be best to place them in one spot. microvm and Hyperlight HALs share most of these constants too.

danbugs committed:
[libs] E: added symbolic link to config-files dir in config lib

Signed-off-by: danbugs <danilochiarlone@gmail.com>
